pkg/cosigner/image_signer.go

```diff
@@ -0,0 +1,101 @@
+package cosigner
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+
+	"github.com/buildpacks/lifecycle"
+	"github.com/sigstore/cosign/cmd/cosign/cli"
+)
+
+type SignFunc func(ctx context.Context, ko cli.KeyOpts, annotations map[string]interface{}, imageRef string, certPath string, upload bool, payloadPath string, force bool, recursive bool) error
+
+type ImageSigner struct {
+	Logger   *log.Logger
+	signFunc SignFunc
+}
+
+const (
+	cosignRepositoryEnv       = "COSIGN_REPOSITORY"
+	cosignDockerMediaTypesEnv = "COSIGN_DOCKER_MEDIA_TYPES"
+)
+
+var (
+	secretLocation = "/var/build-secrets/cosign"
+)
+
+func NewImageSigner(logger *log.Logger, signFunc SignFunc) *ImageSigner {
+	return &ImageSigner{
+		Logger:   logger,
+		signFunc: signFunc,
+	}
+}
+
+func (s *ImageSigner) Sign(ctx context.Context, report lifecycle.ExportReport, annotations map[string]interface{}, cosignRepositories map[string]interface{}, cosignDockerMediaTypes map[string]interface{}) error {
+	cosignSecrets, err := findCosignSecrets()
+	if err != nil {
+		s.Logger.Printf("no keys found for cosign signing: %v\n", err)
+		return nil
+	}
+
+	if len(cosignSecrets) == 0 {
+		s.Logger.Println("no keys found for cosign signing")
+		return nil
+	}
+
+	refImage := report.Image.Tags[0]
+
+	for _, cosignSecret := range cosignSecrets {
+		cosignKeyFile := fmt.Sprintf("%s/%s/cosign.key", secretLocation, cosignSecret)
+		cosignPasswordFile := fmt.Sprintf("%s/%s/cosign.password", secretLocation, cosignSecret)
+
+		ko := cli.KeyOpts{KeyRef: cosignKeyFile, PassFunc: func(bool) ([]byte, error) {
+			content, err := ioutil.ReadFile(cosignPasswordFile)
+			if err != nil {
+				return []byte(""), nil
+			}
+
+			return content, nil
+		}}
+
+		if cosignRepository, ok := cosignRepositories[cosignSecret]; ok {
+			os.Setenv(cosignRepositoryEnv, fmt.Sprintf("%s", cosignRepository))
```

Comment:
is there a reason we don't handle the `os.Setenv` errors?

Reply:
I'll update to handle

Reply:
Trying to write the tests for this now

Reply:
it may be hard to test this behavior, I wouldn't make a specific test for these error paths

Reply:
I think it is possible to test if I change the `cosignRepositoryEnv` to be a variable. Then we would override it to "something=" which is invalid and would fail. Then we can validate these error messages.

We will give it a try for a bit. if it fails, we can possibly ignore because it is very unlikely to occur

Reply:
@tylerphelan according to the [Linux manual page on `setenv()`](https://man7.org/linux/man-pages/man3/setenv.3.html), this should only fail when the variable names contain `=` or is empty/null. Would you say it's fair to not have specific tests to check the error path here?
We currently see very little value in writing a test, since we would have to force the flow by changing the environment variable name, which in this case should just be a constant.

Reply:
> Would you say it's fair to not have specific tests to check the error path here?

yea, I don't think testing the behavior of `os.Setenv` is worthwhile